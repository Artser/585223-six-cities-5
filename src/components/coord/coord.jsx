import React, {PureComponent} from "react";
import leaflet from "leaflet";
import PropTypes from "prop-types";
import {coordType, cityType, offerType} from "../../types";
import {connect} from "react-redux";
import {getActiveCity} from "../../reducer/reselect";


const zoom = 12;
const icon = leaflet.icon({
  iconUrl: `img/pin.svg`,
  iconSize: [30, 30]
});

const activeIcon = leaflet.icon({
  iconUrl: `img/pin-active.svg`,
  iconSize: [30, 30]
});

const style = {
  height: `100%`,
  width: `100%`,
};


class Coord extends PureComponent {
  constructor(props) {

    super(props);
    this._map = null;

  }

  _drawActivePin(coord) {
    leaflet
      .marker(coord, {icon: activeIcon, zIndexOffset: 1000})
      .addTo(this._map);
  }

  _resetMap() {
    this._map.remove();
  }

  _setMap() {

    const {activeCity, activeOffer} = this.props;
    this._map = leaflet.map(`map`, {
      center: activeCity.coord,
      zoom,
      zoomControl: false,
      marker: true
    });
    this._map.setView(activeCity.coord, zoom);

    leaflet
      .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
        attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
      })
      .addTo(this._map);
    const {coords} = this.props;
    coords.forEach((coord) => {
      leaflet
        .marker(coord, {icon})
        .addTo(this._map);
    });

    if (activeOffer) {
      this._drawActivePin(activeOffer);
    }
  }


  componentDidUpdate() {
    this._resetMap();
    this._setMap();
  }

  componentDidMount() {
    this._setMap();
  }

  render() {

    return (
      <div id="map" style={style}></div>
    );

  }
}
Coord.propTypes = {
  coords: PropTypes.arrayOf(coordType),
  activeCity: cityType,
  offer: offerType,
  activeOffer: PropTypes.array
};

const mapStateToProps = (state) => {
  return {
    activeCity: getActiveCity(state),
  };
};
export default connect(mapStateToProps)(Coord);


