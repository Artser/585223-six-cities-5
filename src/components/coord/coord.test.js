import React from "react";
import renderer from "react-test-renderer";
import {Coord} from "./coord";

describe(`Render Coord`, () => {
  const offer = {
    adults: 3,
    avatar: `img/avatar-angelina.jpg`,
    bedrooms: 1,
    city: `Amsterdam`,
    cityId: 4,
    coord: [52.388540000000006, 4.899976],
    description: `A new spacious villa, one floor.`,
    id: 5,
    images: [`https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`],
    imgLink: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    isFavorite: true,
    isPremium: false,
    name: `Angelina`,
    price: 134,
    type: `test`,
    rating: 2.2,
    goods: [test],
    threeImages: [`https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`],
    title: `The house among olive`
  };
  const coords = [[42, 54]];
  const activeCity = {
    name: `Amsterdam`,
    coord: [54, 45],
    id: 2
  };

  it(`Render Coord`, () => {

    const tree = renderer
      .create(
          <Coord
            coords={coords}
            activeCity={activeCity}
            activeOffer={[offer]}
            offer={offer}
          />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
