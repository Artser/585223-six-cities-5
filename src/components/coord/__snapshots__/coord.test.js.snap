// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Coord Render Coord 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;
